File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/spec/config/options_spec.rb
# frozen_string_literal: true

require "spec_helper"

RSpec.describe GreySdk::Config::Options do
  describe ".new" do
    it "creates options with required parameters" do
      options = described_class.new(host: "api.example.com")

      expect(options.host).to eq("api.example.com")
      expect(options.port).to eq(443)
      expect(options.use_tls).to be true
      expect(options.timeout).to eq(30)
      expect(options.headers).to eq({})
    end

    it "creates options with all parameters" do
      options = described_class.new(
        host: "localhost",
        port: 8080,
        use_tls: false,
        timeout: 60,
        headers: { "X-Custom" => "value" }
      )

      expect(options.host).to eq("localhost")
      expect(options.port).to eq(8080)
      expect(options.use_tls).to be false
      expect(options.timeout).to eq(60)
      expect(options.headers).to eq({ "X-Custom" => "value" })
    end
  end

  describe ".local" do
    it "creates options for local development" do
      options = described_class.local(3000)

      expect(options.host).to eq("localhost")
      expect(options.port).to eq(3000)
      expect(options.use_tls).to be false
    end

    it "uses default port 8080" do
      options = described_class.local

      expect(options.port).to eq(8080)
    end
  end

  describe ".production" do
    it "creates options for production" do
      options = described_class.production("api.grey.com")

      expect(options.host).to eq("api.grey.com")
      expect(options.port).to eq(443)
      expect(options.use_tls).to be true
    end

    it "allows custom port" do
      options = described_class.production("api.grey.com", 8443)

      expect(options.port).to eq(8443)
    end
  end

  describe "#base_url" do
    it "returns https URL for TLS" do
      options = described_class.production("api.grey.com")

      expect(options.base_url).to eq("https://api.grey.com:443")
    end

    it "returns http URL without TLS" do
      options = described_class.local(8080)

      expect(options.base_url).to eq("http://localhost:8080")
    end
  end

  describe "#with_timeout" do
    it "returns new options with different timeout" do
      original = described_class.local
      modified = original.with_timeout(120)

      expect(original.timeout).to eq(30)
      expect(modified.timeout).to eq(120)
      expect(modified.host).to eq(original.host)
    end
  end

  describe "#with_headers" do
    it "returns new options with merged headers" do
      original = described_class.new(host: "api.grey.com", headers: { "X-A" => "a" })
      modified = original.with_headers({ "X-B" => "b" })

      expect(original.headers).to eq({ "X-A" => "a" })
      expect(modified.headers).to eq({ "X-A" => "a", "X-B" => "b" })
    end
  end

  describe "#to_h" do
    it "converts options to hash" do
      options = described_class.local

      expect(options.to_h).to include(
        host: "localhost",
        port: 8080,
        use_tls: false
      )
    end
  end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-haskell/test/OptionsSpec.hs
{-# LANGUAGE OverloadedStrings #-}

module OptionsSpec (spec) where

import Test.Hspec
import qualified Data.Map.Strict as Map
import Grey.Config.Options

spec :: Spec
spec = do
  describe "defaultOptions" $ do
    it "has localhost as host" $ do
      optHost defaultOptions `shouldBe` "localhost"

    it "has port 50051" $ do
      optPort defaultOptions `shouldBe` 50051

    it "has TLS disabled" $ do
      optUseTls defaultOptions `shouldBe` False

    it "has 30 second timeout" $ do
      optTimeoutSecs defaultOptions `shouldBe` 30

    it "has no auth token" $ do
      optAuthToken defaultOptions `shouldBe` Nothing

    it "has empty custom headers" $ do
      optCustomHeaders defaultOptions `shouldBe` Map.empty

  describe "localOptions" $ do
    it "uses default port when Nothing" $ do
      optPort (localOptions Nothing) `shouldBe` 50051

    it "uses custom port when provided" $ do
      optPort (localOptions (Just 9000)) `shouldBe` 9000

    it "has TLS disabled" $ do
      optUseTls (localOptions Nothing) `shouldBe` False

  describe "productionOptions" $ do
    it "uses provided host" $ do
      optHost (productionOptions "api.example.com" Nothing) `shouldBe` "api.example.com"

    it "uses default port 443 when Nothing" $ do
      optPort (productionOptions "api.example.com" Nothing) `shouldBe` 443

    it "uses custom port when provided" $ do
      optPort (productionOptions "api.example.com" (Just 8443)) `shouldBe` 8443

    it "has TLS enabled" $ do
      optUseTls (productionOptions "api.example.com" Nothing) `shouldBe` True

  describe "withHost" $ do
    it "updates host" $ do
      let opts = withHost "newhost" defaultOptions
      optHost opts `shouldBe` "newhost"

    it "preserves other fields" $ do
      let opts = withHost "newhost" defaultOptions
      optPort opts `shouldBe` 50051

  describe "withPort" $ do
    it "updates port" $ do
      let opts = withPort 9000 defaultOptions
      optPort opts `shouldBe` 9000

  describe "withTls" $ do
    it "enables TLS" $ do
      let opts = withTls True defaultOptions
      optUseTls opts `shouldBe` True

    it "disables TLS" $ do
      let opts = withTls False (productionOptions "host" Nothing)
      optUseTls opts `shouldBe` False

  describe "withTimeout" $ do
    it "updates timeout" $ do
      let opts = withTimeout 60 defaultOptions
      optTimeoutSecs opts `shouldBe` 60

  describe "withAuthToken" $ do
    it "sets auth token" $ do
      let opts = withAuthToken "token123" defaultOptions
      optAuthToken opts `shouldBe` Just "token123"

  describe "withHeader" $ do
    it "adds custom header" $ do
      let opts = withHeader "X-Custom" "value" defaultOptions
      Map.lookup "X-Custom" (optCustomHeaders opts) `shouldBe` Just "value"

    it "can add multiple headers" $ do
      let opts = withHeader "X-One" "1" $ withHeader "X-Two" "2" defaultOptions
      Map.lookup "X-One" (optCustomHeaders opts) `shouldBe` Just "1"
      Map.lookup "X-Two" (optCustomHeaders opts) `shouldBe` Just "2"

  describe "getEndpoint" $ do
    it "returns host:port format" $ do
      getEndpoint defaultOptions `shouldBe` "localhost:50051"

    it "works with production options" $ do
      getEndpoint (productionOptions "api.example.com" Nothing) `shouldBe` "api.example.com:443"
